File: go/utils/math.go
package utils

import (
	"math"

	"golang.org/x/exp/constraints"
)

type Number interface {
	constraints.Integer | constraints.Float
}

// Clamp returns num clamped to [low, high]
func Clamp[T Number](num T, low T, high T) T {
	if num < low {
		return low
	}
	if num > high {
		return high
	}
	return num
}

// Map from [sourceLow, sourceHigh] to [targetLow, targetHigh] linearly
func MapNumber[T Number](num T, sourceLow T, sourceHigh T, targetLow T, targetHigh T) T {
	var mapped = ((num-sourceLow)*(targetHigh-targetLow))/(sourceHigh-sourceLow) + targetLow
	return Clamp(mapped, targetLow, targetHigh)
}

func ProbToOdds(prob float64) float64 {
	return math.Log(prob / (1 - prob))
}

func OddsToProb(odds float64) float64 {
	return math.Exp(odds) / (1 + math.Exp(odds))
}

File: src/utils/math.ts
/**
 * Maps a number from one range to another
 * @param value The value to map
 * @param inMin The minimum value of the input range
 * @param inMax The maximum value of the input range
 * @param outMin The minimum value of the output range
 * @param outMax The maximum value of the output range
 * @returns The mapped value
 */
export function mapNumber(
  value: number,
  inMin: number,
  inMax: number,
  outMin: number,
  outMax: number,
): number {
  return ((value - inMin) * (outMax - outMin)) / (inMax - inMin) + outMin;
}

export function probToOdds(prob: number): number {
  return prob / (1 - prob);
}

export function oddsToProb(odds: number): number {
  return odds / (1 + odds);
}
